docs/compatibility_draft.md

```diff
@@ -0,0 +1,246 @@
+- Downgrade IR to shim before bytecoding `arith.addi → stablehlo.arith.addi`
+- Upgrade IR to other dialect on deserialization.
+
+With the above rules, we are able to provide compatibility guarantees for Migration and Additions. 
+
+_Example:_ Move `builtin.func → func.func`.
+
+<img width="943" alt="dependent" src="https://user-images.githubusercontent.com/3036970/189246267-06991d87-71a7-4ea8-bca0-5c818b0f2964.png">
+
+There are still some open questions that require more discussion, regarding semantic changes or breaking changes to ops in other dialects. One option would be to do a deeper copy of the op into StableHLO and support it in our interpreter and transformations.
+
+## Enumerating Possible Changes - Validating the Design
+This section enumerates different changes that are possible in a dialect. The goal of this enumeration is to stress test the proposed protocols in “Compatibility Protocols”.
+
+<details>
+<summary>Expand enumeration of possible changes.</summary>
+The following changes are possible ways that a dialect can evolve:
+
+**StableHLO Operations**
+- Add Operation - Addition change
+- Remove Opeartion - Deprecation change
+- Rename Operation - Compatible change
+- Rename Operation Mnemonic - Migration change
+- Adding an operand to an op - Versioned op change
+- Adding a result to an op - Versioned op change
+
+**StableHLO Attributes / Types / Enums**
+- Rename attribute / type / enum - Migration change
+- Add attribute / type / enum - Addition change
+- Remove an attribute / type / enum from op - Deprecation change
+- Change the type of an attribute - Migration change if upgrade/downgrad possible, else Versioned op change.
+- Changing type signature - ???
+- Changing an attribute into an operand - Migration change if upgrade/downgrad possible, else Versioned op change. 
+
+**Other StableHLO Features**
+- Changing verifier - Compatible change if bug fix / UB-to-error, else versioned op change
+- Changing type inference - Same as verifier.
+- Changing prettyprinter - Compatible change
+- Changing folders / canonicalizers - ??
+
+**Changes in Dependent Dialects** - This needs more thought. Currently migration changes and addition changes are easily supported. A follow-up proposal will investigate this further.
+</details>
+
+## Detecting Incompatibilities (Test Plan)
+
+### Detecting Backward Incompatibility
+An approach similar to [LLVM bitcode compatibility testing](https://llvm.org/docs/DeveloperPolicy.html#ir-backwards-compatibility) can be used for StableHLO:
+> “After each X.Y release, compatibility.ll must be copied to compatibility-X.Y.ll. The corresponding bitcode file should be assembled using the X.Y build and committed as compatibility-X.Y.ll.bc.”
+
+Currently, we have large integ test files for StableHLO (`ops_stablehlo_roundtrip.mlir`) and CHLO (`ops_chlo_roundtrip.mlir`). Backwards compatibility testing can be completed locally and in the “Build and Test” GitHub Actions using the following procedure:
```

Comment:
Typo: "integration".

Reply:
Fixed this and greatly re-worded the testing section to include mention of dynamic testing using the interpreter to ensure semantics are guaranteed across versions.